File: superfriend-py/src/superfriend/services/agent_session_manager.py
from __future__ import annotations

import time
import structlog
from enum import Enum
from dataclasses import dataclass, field
from typing import Any

logger = structlog.get_logger()


class SessionState(str, Enum):
    IDLE = "idle"
    EXECUTING = "executing"
    PAUSED = "paused"
    CANCELLED = "cancelled"


class InterruptMode(str, Enum):
    CANCEL = "cancel"
    APPEND = "append"


@dataclass
class AgentSessionState:
    state: SessionState = SessionState.IDLE
    interrupt_mode: InterruptMode | None = None
    appended_context: str | None = None
    start_time: float = 0.0


class AgentSessionManager:
    def __init__(self):
        self._active_sessions: dict[str, AgentSessionState] = {}

    def start_execution(self, session_id: str) -> None:
        state = self._active_sessions.setdefault(session_id, AgentSessionState())
        state.state = SessionState.EXECUTING
        state.interrupt_mode = None
        state.appended_context = None
        state.start_time = time.time()
        logger.info("agent_session_started", session_id=session_id)

    def end_execution(self, session_id: str) -> None:
        removed = self._active_sessions.pop(session_id, None)
        if removed:
            duration = time.time() - removed.start_time
            logger.info("agent_session_ended", session_id=session_id, duration_ms=int(duration * 1000))

    def should_interrupt(self, session_id: str) -> bool:
        state = self._active_sessions.get(session_id)
        return state is not None and state.interrupt_mode is not None

    def get_interrupt_mode(self, session_id: str) -> InterruptMode | None:
        state = self._active_sessions.get(session_id)
        return state.interrupt_mode if state else None

    def get_appended_context(self, session_id: str) -> str | None:
        state = self._active_sessions.get(session_id)
        return state.appended_context if state else None

    def interrupt(self, session_id: str, mode: InterruptMode, context: str | None = None) -> None:
        state = self._active_sessions.get(session_id)
        if state is None:
            logger.warning("cannot_interrupt_session_not_found", session_id=session_id)
            return
        if state.state != SessionState.EXECUTING:
            logger.warning("cannot_interrupt_wrong_state", session_id=session_id, state=state.state)
            return
        state.interrupt_mode = mode
        state.appended_context = context
        if mode == InterruptMode.CANCEL:
            state.state = SessionState.CANCELLED
        else:
            state.state = SessionState.PAUSED
        logger.info("session_interrupted", session_id=session_id, mode=mode.value)

    def pause(self, session_id: str) -> None:
        state = self._active_sessions.get(session_id)
        if state and state.state == SessionState.EXECUTING:
            state.state = SessionState.PAUSED
            logger.info("session_paused", session_id=session_id)

    def resume(self, session_id: str) -> None:
        state = self._active_sessions.get(session_id)
        if state and state.state == SessionState.PAUSED:
            state.state = SessionState.EXECUTING
            state.interrupt_mode = None
            state.appended_context = None
            logger.info("session_resumed", session_id=session_id)

    def get_session_state(self, session_id: str) -> SessionState | None:
        state = self._active_sessions.get(session_id)
        return state.state if state else None

    def is_executing(self, session_id: str) -> bool:
        state = self._active_sessions.get(session_id)
        return state is not None and state.state == SessionState.EXECUTING

    def get_active_sessions(self) -> list[str]:
        return [
            sid for sid, state in self._active_sessions.items()
            if state.state == SessionState.EXECUTING
        ]

    def get_all_sessions(self) -> dict[str, dict[str, Any]]:
        return {
            sid: {
                "state": state.state.value,
                "interrupt_mode": state.interrupt_mode.value if state.interrupt_mode else None,
                "start_time": state.start_time,
                "duration_ms": int((time.time() - state.start_time) * 1000) if state.start_time else 0,
            }
            for sid, state in self._active_sessions.items()
        }


agent_session_manager = AgentSessionManager()

File: src/main/java/com/superfriend/superfriend/service/AgentSessionManager.java
package com.superfriend.superfriend.service;

import lombok.Data;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

//会话状态管理
@Slf4j
@Service
public class AgentSessionManager {

    private final ConcurrentHashMap<String, AgentSessionState> activeSessions = new ConcurrentHashMap<>();

    public enum SessionState {
        IDLE,
        EXECUTING,
        PAUSED,
        CANCELLED
    }

    public enum InterruptMode {
        CANCEL,
        APPEND
    }

    @Data
    public static class AgentSessionState {
        private SessionState state = SessionState.IDLE;
        private InterruptMode interruptMode;
        private String appendedContext;
        private long startTime;
    }

    public void startExecution(String sessionId) {
        AgentSessionState state = activeSessions.computeIfAbsent(sessionId, k -> new AgentSessionState());
        state.setState(SessionState.EXECUTING);
        state.setInterruptMode(null);
        state.setAppendedContext(null);
        state.setStartTime(System.currentTimeMillis());
        log.info("Agent session started: {}", sessionId);
    }

    public void endExecution(String sessionId) {
        AgentSessionState removed = activeSessions.remove(sessionId);
        if (removed != null) {
            long duration = System.currentTimeMillis() - removed.getStartTime();
            log.info("Agent session ended: {}, duration: {}ms", sessionId, duration);
        }
    }

    public boolean shouldInterrupt(String sessionId) {
        AgentSessionState state = activeSessions.get(sessionId);
        return state != null && state.getInterruptMode() != null;
    }

    public InterruptMode getInterruptMode(String sessionId) {
        AgentSessionState state = activeSessions.get(sessionId);
        return state != null ? state.getInterruptMode() : null;
    }

    public String getAppendedContext(String sessionId) {
        AgentSessionState state = activeSessions.get(sessionId);
        return state != null ? state.getAppendedContext() : null;
    }

    public void interrupt(String sessionId, InterruptMode mode, String context) {
        AgentSessionState state = activeSessions.get(sessionId);
        if (state == null) {
            log.warn("Cannot interrupt session {}: session not found", sessionId);
            return;
        }
        if (state.getState() != SessionState.EXECUTING) {
            log.warn("Cannot interrupt session {}: current state is {}", sessionId, state.getState());
            return;
        }
        state.setInterruptMode(mode);
        state.setAppendedContext(context);
        log.info("Agent session interrupt requested: {}, mode={}, context={}", sessionId, mode, context);
    }

    public void clearInterrupt(String sessionId) {
        AgentSessionState state = activeSessions.get(sessionId);
        if (state != null) {
            state.setInterruptMode(null);
            state.setAppendedContext(null);
        }
    }

    public SessionState getSessionState(String sessionId) {
        AgentSessionState state = activeSessions.get(sessionId);
        return state != null ? state.getState() : SessionState.IDLE;
    }

    public boolean isExecuting(String sessionId) {
        return getSessionState(sessionId) == SessionState.EXECUTING;
    }

    /**
     * 取消正在执行的会话
     * @param sessionId 会话ID
     * @return 是否成功取消
     */
    public boolean cancelSession(String sessionId) {
        AgentSessionState state = activeSessions.get(sessionId);
        if (state == null) {
            log.warn("Cannot cancel session {}: session not found", sessionId);
            return false;
        }
        if (state.getState() != SessionState.EXECUTING) {
            log.warn("Cannot cancel session {}: current state is {}", sessionId, state.getState());
            return false;
        }
        state.setState(SessionState.CANCELLED);
        state.setInterruptMode(InterruptMode.CANCEL);
        log.info("Agent session cancelled: {}", sessionId);
        return true;
    }

    /**
     * 检查会话是否已被取消
     * @param sessionId 会话ID
     * @return 是否已取消
     */
    public boolean isCancelled(String sessionId) {
        AgentSessionState state = activeSessions.get(sessionId);
        return state != null &&
               (state.getState() == SessionState.CANCELLED ||
                state.getInterruptMode() == InterruptMode.CANCEL);
    }

    /**
     * 获取所有活跃会话数量
     */
    public int getActiveSessionCount() {
        return (int) activeSessions.values().stream()
            .filter(s -> s.getState() == SessionState.EXECUTING)
            .count();
    }

    /**
     * 清理超时的会话（超过指定毫秒数）
     */
    public int cleanupStaleSessions(long timeoutMs) {
        long now = System.currentTimeMillis();
        int cleaned = 0;
        for (Map.Entry<String, AgentSessionState> entry : activeSessions.entrySet()) {
            if (entry.getValue().getState() == SessionState.EXECUTING &&
                now - entry.getValue().getStartTime() > timeoutMs) {
                activeSessions.remove(entry.getKey());
                cleaned++;
                log.warn("Cleaned up stale session: {}, age: {}ms",
                    entry.getKey(), now - entry.getValue().getStartTime());
            }
        }
        return cleaned;
    }
}
